docs/tutorial.rst

```diff
@@ -1224,65 +1224,54 @@ of at least 0.8 by the end. Next, we'll walk through the steps required to do th
     this conditioning works by re-running the simulation until the condition is
     achieved, a nearly impossible condition will result in very long run times.
 
-First, we need to define the mutation type for the selected mutation.
-So we can refer to it later, we need its "mutation type ID". This is just
-the index of the new mutation type in the contig's list of mutation types.
+First, we need to add a single site mutation in the middle of the contig.
+This represents a mutation of a single base pair, with associated fitness and
+dominance coefficients.  The single site mutation must have a unique string ID
+so that we can refer to it later.
 
 .. code-block:: python
 
-    mt = stdpopsim.MutationType(
-        distribution_type="f",
-        dominance_coeff=1.0,
-        distribution_args=[0.5],
-        convert_to_substitution=False,
-    )
-    dfe = stdpopsim.DFE(
-        id="new_mutation",
-        mutation_types=[mt],
-        proportions=[1.0],
-        description="added mutation",
-        long_description="mutation type to be added",
-    )
-    contig.add_dfe(
-        intervals=np.empty((0, 2), dtype="int"),
-        DFE=dfe,
+    mut_id = "hard sweep"
+    coordinate = round(contig.recombination_map.sequence_length / 2)
+    contig.add_single_site(
+        id=mut_id,
+        coordinate=coordinate,
+        fitness_coeff=0.1,
+        dominance_coeff=0.5,
     )
-    for mt_info in contig.mutation_types():
-        if mt_info["dfe_id"] == dfe.id:
-            break
 
-    mut_id = mt_info["id"]
+.. note::
+
+    Note that single site mutations are internally stored as DFEs, and DFEs
+    cannot overlap. As a consequence, if a DFE is added to the contig over an
+    interval that contains an existing single site mutation, the latter will be
+    "overwritten" and an error will be raised by the simulation engine.
 
 Next, we will set up the "extended events" which will modify the demography.
-The first extended event is placing of the selected mutation,
-which will occur in a random individual from the first population (id 0),
-in the middle of the contig, 1000 generations ago.
-We specify ``save=True`` to ``stdpopsim.ext.DrawMutation``
-so that the simulation can restart from that point if the mutation is lost.
+The first extended event is the origination of the selected mutation, which
+will occur in a random individual from the first population (id 0), 1000
+generations ago.  We specify ``save=True`` to ``stdpopsim.ext.DrawMutation`` so
+that the simulation can restart from that point if the mutation is lost.
 
 .. code-block:: python
 
-    coordinate = round(contig.recombination_map.sequence_length / 2)
     T_mut = 1000
     extended_events = [
         stdpopsim.ext.DrawMutation(
             time=T_mut,
-            mutation_type_id=mut_id,
+            single_site_id=mut_id,
             population_id=0,
-            coordinate=coordinate,
             save=True,
```

Comment:
Shouldn't `save` always be `True`?  When would we not want it to be?

Reply:
Then I'll go ahead and remove this option; on the slim side of things, too.

Reply:
We should always save state for allele frequency conditioning, right? Currently there is a `save` option in `ext.ConditionOnAlleleFrequency`, that I can also remove

Reply:
Well, there's a disk space/speed trade-off here: I think the way it works is that when it does the conditioning it jumps back to the most recently saved time; if it's saved more often that will be more efficient. However, it seems like unnecessary complexity for the API here - just saving all the time should be fine, unless someone is doing some overly complicated conditioning. I guess it's worth checking, though - since conditioning can act over and *interval* of time (right?), at what point in that interval does the save happen?

Reply:
The save occurs just once, at the beginning of the interval. There is a speed cost associated with saving because of writing out the tree sequence, but I don't think there should be extra use of disk space -- because the slim function `save` is always writing to the same file (the global `trees_file`) ... unless I'm missing something? With complicated conditioning, it seems to me that the rejection sampling would have difficultly completing *without* saving frequently. But, I'm happy to punt this to another issue if there's any question about removing the `save` toggle from `ext.ConditionOnAlleleFrequency`.

Reply:
Ok, we've looked at the SLiM code and we think that there's not a big enough benefit to providing the flexibility to optionally save or not save at these conditioning events to warrant the more complicated interface. Does this seem OK to you, @grahamgower? (tldr; we want to remove the `save=` option from the API for these events, since it seems like unnecessary complexity)

Reply:
> With complicated conditioning, it seems to me that the rejection sampling would have difficultly completing _without_ saving frequently.

Yes, this was the main reason for having the `save=` option. The initial selection implementation was strongly motivated by the desire to model adaptive introgression as in https://github.com/popsim-consortium/stdpopsim/blob/main/docs/selection_example.py (and be able to also do this on top of a DFE).

I'm not sure if there's a simpler way to achieve this?

Reply:
I guess that example should still work with `save=True` in all conditioning events. So probably its ok to remove the option and always do that. I almost certainly chose to store prior to the introduction of the selected mutation because I was following a SLiM recipe (and had about 5 mins experience with SLiM at the time), so it would be a good idea to also check the standard SLiM recipe(s) to see if there's any reasoning that we're missing here.

Reply:
That makes perfect sense to me @grahamgower -- what we were thinking was to always save the state when a mutation is drawn or at the beginning of a frequency conditioning interval; and drop the `save` toggle from the API. Does this seem OK, or can you think of a case when `save=False` would make sense?

Reply:
Sounds good!

Reply:
Thinking about this a bit more... if one conditions on the allele not being lost and saves the simulation state, the allele could be at very low frequency; if one waits `T` generations then conditions on the allele at high frequency, then the chance of the allele having rised in frequency when `T` is small is negligible and so the simulation enters into an endless save/restore loop. Clearly one must be careful about AF conditioning. But, having the option for `save=False` allows one to change the AF conditioning, without having to worry about this problem. This might be desirable when `T` is not known in advance, but is drawn from a distribution (as I was doing to generate training data for a CNN).

Reply:
Currently, conditioning events have fixed start/end times and require the associated `DrawMutation` event to have `save=True` -- so simulations would still end up in an endless restore cycle if the AF condition is unrealistic. To avoid this in the way you describe, the random AF conditioning start time would have to be drawn on the SLiM side of things, if I'm understanding correctly? Alternatively, there could be an option forcing SLiM to error out if a simulation fails an AF condition, so that a rejection sampler w/ random times could be implemented via a `try` block on the python side. Either way, sounds like enough infrastructure to warrant a separate PR.

Reply:
The complete working code is here: https://github.com/grahamgower/genomatnn/blob/main/genomatnn/sim.py#L645

I've needed to have `genomatnn` depend on a stdpopsim branch/fork which is pinned at a post-stdpopsim-0.1.2 version that worked when I wrote this, as I was uncertain what the final stdpopsim API would look like when we finally push out the next stdpopsim release. If the `save=` option is removed I would have a hard time updating the genomatnn code to use the next stdpopsim release. Although this wouldn't be the end of the world, as updating that code isn't a high priority for me anyhow.

Reply:
Ah, I think I understand, but let me spell it out to be sure. Say you're simulating a frequency trajectory $f$ from generation $t$ until the present, conditional on lower bounds $x_a, x_b$ at times $b > a > t$. So, drawing a sample from $$\mathrm{Pr}(f_{t ... 0} \mid f_{a...b-1} > x_a, f_{b...0} > x_b) = \mathrm{Pr}(f_{t ... a-1}) \times \mathrm{Pr}(f_{a ... b-1} \mid f_{a...b-1} > x_a, f_{a-1}) \times \mathrm{Pr}(f_{b ... 0} \mid f_{b...0} > x_b, f_{b-1})$$ If `save=True` universally, then we're rejection sampling sequentially from each of the terms on the right hand side. The problem is that the rejection rate might effectively be 1 when sampling from $\mathrm{Pr}(f_{b ... 0} \mid f_{b...0} > x_b, f_{b-1})$ if $f_{b-1} \ll x_b$. So instead of sampling from $\mathrm{Pr}(f_{a ... b-1} \mid f_{a...b-1} > x_a, f_{a-1})$ followed by $\mathrm{Pr}(f_{b ... 0} \mid f_{b ... 0} > x_b, f_{b-1})$, it might be much more efficient to just rejection sample once from the joint distribution $\mathrm{Pr}(f_{a ... b-1}, f_{b ... 0} \mid f_{a ... b-1} > x_a, f_{b ... 0} > x_b, f_{a-1})$ -- in fact this might be the only way to get the sampler to reliably finish in a reasonable amount of time.

This does seem useful -- especially if we're ultimately going to de-emphasize this lower-level API in favor of convenience functions for specific models (so that more complexity in the lower-level interface isn't so much an issue). So, how about we keep `save` for `ConditionOnAlleleFrequency` and get rid of `save` for `DrawMutation`. What do you think, @petrelharp and @andrewkern?

Reply:
Yeah this is a good argument in favor of keeping `save` with `ConditionOnAlleleFrequency` -- agreed. 

Sounds good to me @nspope

Reply:
Um, hold on here - maybe this is what you're saying already, but: even though $f$ is Markov,
$$
  \mathbb{P}(f_t > x | f_{s_i} > x_i \text{ for } 1 \le i \le k)
  \neq
  \mathbb{P}(f_{s_1} > x_1) \mathbb{P}(f_{s_2} > x_2 | f_{s_1} > x_1} \cdots \mathbb{P}(f_t > x | f_{s_k} > x_k} ,
$$
and so besides any issue of efficiency of the rejection sampling, there's actually an issue of *correctness* of the conditioning. To do the conditioning correctly, we've got to restart all the way back to the first event.

So, I think that we need `save=True` for when the mutation is introduced, but we should *not* have it for any subsequent events (unfortunately)? I'd want to think through this more carefully to be sure, but I currently don't see how doing `save=True` at any subsequent events will actually give you a correctly conditioned simulation.

Reply:
Maybe I'm not following, but surely this is ok if we first condition on AF>0, then condition on (say) AF>0.2. There's no other way to obtain AF>0.2 except if AF>0 in every proceeding generation.

Reply:
I might be being dense here, @petrelharp -- but if $f$ is Markov, then 
$$\mathrm{Pr}(f_{i}\~\mathrm{for}\~i\~\mathrm{in}\~1 \leq i \leq k) = \mathrm{Pr}(f_{1}) \times \mathrm{Pr}(f_{2} \mid f_{1}) \times \cdots \times \mathrm{Pr}(f_{k} \mid f_{k-1})$$ And,
$$\mathrm{Pr}(f_{1 \leq i \leq k} \mid f_{i} > x_i\~\mathrm{for}\~i\~\mathrm{in}\~1 \leq i \leq k) = \mathrm{Pr}(f_{1}) \mathbb{I}[f_{1} > x_1] \times \mathrm{Pr}(f_{2} \mid f_{1}) \mathbb{I}[f_{2} > x_2] \times \cdots \times \mathrm{Pr}(f_{k} \mid f_{k-1}) \mathbb{I}[f_{k} > x_k] $$ So why wouldn't sequentially rejection sampling each of the $\mathrm{Pr}(\cdot \mid \cdot) \mathbb{I}[\cdot]$ terms -- that is, conditioning on $f_i$ for $i < j$ to rejection sample $f_j$ -- produce a draw from the right distribution?

Reply:
Ok, I was totally wrong, you all are right. This is a general fact about stochastic processes, and here's the basic calculation:
![calcs](https://user-images.githubusercontent.com/1046249/183531697-3c27ecda-f75d-49ba-918b-d92356127087.jpg)


In "words", if X is a random variable and Y is X|A, then X|(A&B) and Y|B have the same distribution. So, let X be the random frequency trajectory, let A be the event that the frequency at time t1 is in some interval, and B be the event that the frequency at some later time t2 is in some other interval -- conditioning first on the A then on B is the same as conditioning on both at the same time.

Reply:
In conclusion, `save=True` is correct in multiple `ConditionOnAlleleFrequency`, but may not always be desirable for efficient rejection sampling. However, `save=True` should always be the case for `DrawMutation`. So, we'll remove `save` from `DrawMutation` and retain it for `ConditionOnAlleleFrequency`. For the future, higher-level interface for sweeps in `stdpopsim`, the models will be simple enough that I don't think we'll have to expose `save`.

Reply:
Thanks for spending the time to think this all through @petrelharp and @nspope!

The only reason I can think of to want to set `save=False` in `DrawMutation()` is if we one day support multiple selected mutations in draw mutation. E.g. a drawn mutation in each of two different populations, potentially drawn at different times. A use case like this ought not take much work to support (probably just fix the AF calculation in the SLiM code). Is removing `save` from DrawMutation a big win in terms of code simplification?

Reply:
I'm ambivalent @grahamgower -- I think the eventual goal is to have a higher level interface for simple sweep scenarios that is cleaner, but that produces lists of these extended event classes under the hood. If you see an advantage to keeping `save` arguments in all the events (in terms of flexibility and future extensions and so on), then that's good enough for me! We might want to document some of these nuances, though, in case this comes up again.

Reply:
Um, after some more consideration, I'm worried about the entire rejection sampling set-up? I'm probably missing something silly, but: consider the case where you introduce a mutation at time 0 and condition on it being not lost at both 100 and 200 generations later. As Andy pointed out yesterday, trajectories that are not lost at both times will tend to be higher at time 100 then trajectories that are not lost at just time 100. In other words: let f(t) be the frequency at time t, then
$$\mathbb{P}(f(100) > x | f(100) > 0 \text{ and } f(200) > 0) > \mathbb{P}(f(100) > x | f(100) > 0)$$
However, if we rejection sample back to time 100, then the distribution of frequencies at time 100 will be given by the expression on the right, when we want the expression on the left.

The problem with the argument on the whiteboard was that the rejection sampling isn't actually sampling from that conditional distribution.

And, here's some R code that demonstrates that this intermediate rejection sampling doesn't get the right thing. Consider $Z = X + Y$; say we want to sample from $Z$ given $Z > 1$. This is equivalent to conditioning on $Z$ given $Z > 1$ and $X > -N$ for large enough $N$, which amounts to no conditioning on $N$. So, if our rejection sampling approach worked, then we should be able to (a) first sample $X$, and then (b) repeatedly sample $Y$ until we get $X + Y > 1$. However, that's clearly (in some cases) not the same as repeatedly sampling $X+Y$ until we get $X+Y > 1$.
```
f <- function (n, a, b, x) {
    out <- c()
    while (length(out) < n) {
        z <- a(n) + b(n)
        out <- c(out, z[z > x])
    }
    return(out[1:n])
}

g <- function (n, a, b, x) {
    out <- u <- a(n)
    free <- rep(TRUE, n)
    while (any(free)) {
        u <- out + b(n)
        ut <- free & (u > x)
        out[ut] <- u[ut]
        free[ut] <- FALSE
    }
    return(out)
}

a <- rnorm
b <- rnorm
x <- 1
n <- 1000
fn <- f(n, a, b, x)
gn <- g(n, a, b, x)

layout(1:2)
hist(fn, breaks=40, main=expression(paste("x + y | ", x + y > 1)))
hist(gn, breaks=40, main=expression(paste("x + (y | ", x + y > 1, ")")))
```
which gets
![Screenshot from 2022-08-09 08-44-02](https://user-images.githubusercontent.com/1046249/183697769-8de33ee6-4795-4188-a2a3-8f67d3259b73.png)

Reply:
> $\mathbb{P}(f(100) > x | f(100) > 0 \text{ and } f(200) > 0) > \mathbb{P}(f(100) > x | f(100) > 0)$
> However, if we rejection sample back to time 100, then the distribution of frequencies at time 100 will be given by the expression on the right, when we want the expression on the left.

I'm confused by the setup here. If you take only simulations that finish with $f(200) > 0$, then surely you have the expression on the left? My assumption when rejection sampling is that $\mathbb{P}(f(200) > x | f(100) > 0) = \mathbb{P}(f(200) > 0)$---is there something I'm missing here?

Reply:
I'm confused as well. 

1. **Single rejection step**: Sample a trajectory $f(t)$ from time 0 until time 200, accept the trajectory if $f(100) > 0 \land f(200) > 0$, and store the value of $u = f(100)$. This will be a random sample from $\mathrm{Pr}(f(100) \mid f(100) > 0 \land f(200) > 0)$.
2. **Sequential rejection steps**: 
    - Sample a trajectory $f(t)$ from time 0 until time 100, accept the (incomplete) trajectory if $f(100) > 0$, and store $u = f(100)$. This will be a sample from $\mathrm{Pr}(f(100) \mid f(100) > 0)$. 
    - Then simulate a conditioned trajectory $\tilde{f}(t)$ from time 100 to time 200, conditional on $\tilde{f}(100) = u$. Accept  if $\tilde{f}(200) > 0$, and store $v = \tilde{f}(100)$. This will be a sample from $\mathrm{Pr}(f(100) \mid f(100) > 0 \land f(200) > 0)$. 
    - So the first rejection step generates RVs from a distribution of trajectories with a single condition. The second rejection step filters these original RVs to get new RVs that are from a distribution of trajectories with two conditions.

To try to give a simulated example that more closely reflects this process: let $X \sim \mathrm{Uniform}(-1, 1)$. Let $Y = 0$ with probability 0.5 and $Y \sim \mathrm{Uniform}(0, 1)$ with probability 0.5. Let $Z = X Y$. We want samples from $\mathrm{Pr}(X \mid X > 0 \land Z > 0)$ which is just a uniform distribution over [0, 1]. 

Here is R code using a single rejection step (simulate $X, Y$ and accept if $X > 0 \land XY > 0$, versus sequential rejection steps (e.g. first simulate $X$ and accept if $X > 0$, then simulate $Y$ until $XY > 0$).
```r
f <- function(n, draw_X, draw_Y, lb_X, lb_Z) {
  out <- c()
  while (length(out) < n) {
    X <- draw_X()
    Y <- draw_Y()
    Z <- X * Y
    if (X > lb_X & Z > lb_Z) {
      out <- c(out, Y)
    }
  }
  return(out)
}

g <- function(n, draw_X, draw_Y, lb_X, lb_Z) {
  out <- c()
  while (length(out) < n) {
    X <- draw_X()
    if (X > lb_X) {
      while (TRUE) {
        Y <- draw_Y()
        Z <- X * Y
        if (Z > lb_Z) {
          out <- c(out, Y)
          break
        }
      }
    }
  }
  return(out)
}

lb_X <- lb_Z <- 0
n <- 10000
proposal_X <- function() runif(1, -1, 1)
proposal_Y <- function() if(runif(1, 0, 1) < 0.5) 0 else runif(1, 0, 1)
fn <- f(n, proposal_X, proposal_Y, lb_X, lb_Z)
gn <- g(n, proposal_X, proposal_Y, lb_X, lb_Z)
ecdf <- function(x) data.frame(z=sort(x), p=1:length(x)/length(x))

library(ggplot2)
df <- cbind(rbind(ecdf(fn),ecdf(gn)), 
    Algorithm=rep(c("Reject X&Z simultaneously","Reject X&Z sequentially"),each=n)
)
ggplot(df) + 
  geom_step(aes(x=z, y=p, color=Algorithm)) + 
  xlab("Y") + ylab("ECDF") + theme(legend.position=c(0.30,0.8))
```
<img src="https://user-images.githubusercontent.com/5767783/183741157-75b6cec1-f7e8-4561-866e-9068b1231cba.png" width=50%>

This seems fine ... what am I missing?

Reply:
> If you take only simulations that finish with $f(200) > 0$, then surely you have the expression on the left? 

Well, the problem is that the procedure of (simulate back to t=100 and then rejection sample back to t=100 until you are above 0 at 200) is *not* the same thing as simulating $f$ in the first place. You do indeed end up with only trajectories that stay above zero, but not having the right distribution.

> My assumption when rejection sampling is that $\mathbb{P}(f(200) > x | f(100) > 0) = \mathbb{P}(f(200) > x)$ ---is there something I'm missing here?

I think you mean that $\mathbb{P}(f(200) > x \text{ and } f(100) > 0) = \mathbb{P}(f(200) > x)$? That I do agree with; however, it's not true with the conditional.

And, one reason this is so tricky to reason about is that rejection sampling back to an intermediate time makes it not at all obvious what process we're actually simulating. There's two processes:
1. $f$, conditioned on $f(200) > 0$
2. $f$ run until time 100 and conditioned on $f(100) > 0$; and then after that time point conditioned on $f(200) > 0$.

The reason the second process is different is because the process "doesn't know" about the conditioning at $t=200$ for the first 100 time steps.

Reply:
Also note that you could make this correct by using importance weights - if in a particular simulation you have to rejction sample $n$ times until you get a good one, and give that simulation weight $1/n$, then you *do* get the right distribution.

Reply:
Also also, I am **very** sorry to have not noticed this issue much earlier.

Reply:
After talking it out, I think Peter is totally correct -- my counterexample above really doesn't address the problem. Here's my understanding of the issue:

Let $f_a$ be the allele frequency at time $a$, which occurs with probability $\mathrm{Pr}(f_a)$ when we run the simulator from time 0 to $a$. If we subsequently run the simulator from $a$ until a future time $b$ where we check if $f_b > 0$, then the entire trajectory will get accepted with probability $\mathrm{Pr}(f_b > 0 \mid f_a)$ -- so ultimately, our samples of $f_a$ will have the distribution $$\mathrm{Pr}(f_a \mid f_b > 0) = \frac{\mathrm{Pr}(f_a) \mathrm{Pr}(f_b > 0 \mid f_a)}{\mathrm{Pr}(f_b > 0)}$$ where the denominator is just a normalizing constant to make it integrate/sum to 1.

This will be the case when every candidate trajectory is simulated from the beginning. If we're checkpointing, then if some candidate trajectory doesn't satisfy $f_b > 0$, we'd rerun the simulator from the same value of $f_a$ until we finally get a simulation where $f_b > 0$. Thus, the checkpointed $f_a$ gets accepted with probability 1, because we're forcing the rejection sampler to return a trajectory that includes it! In this case, even though the sampled trajectories satisfy $f_b > 0$, they will produce $f_a$ that are distributed according to $\mathrm{Pr}(f_a)$ instead of the conditional distribution $\mathrm{Pr}(f_a \mid f_b > 0)$ (which is what would be produced by the "un-checkpointed" simulator). So the allele frequency trajectories (and thus genealogies) produced by checkpointing won't come from the correct stochastic process.

Does this make sense @grahamgower? I could do some simple simulations to verify that this is indeed the case, if that would help.

Reply:
here's a simple simulation of a random walk on a lattice that demonstrates the issue with multiple conditioning.

in this case i'm conditioning on the process being at some state `x > 0` at `t1` and/or `x > 0.1` at `t2`. Below I plot the random walk and then a boxplot of the final state of the random walk for an unconditioned process, one which we condition on the state at either of the times, or finally when we condition on the state at both of the times.  

```python
import numpy as np
from matplotlib import pyplot as plt

reps = 500
gens = 1000
t1 = gens / 8
t2 = gens / 2
x = np.random.random((gens,reps)) > 0.5
x = x.astype(np.float16) - 0.5
xc = x.cumsum(axis=0)
cond1 = xc[int(t1),:] > 0
cond2 = xc[int(t2), :] > 0.1
cond_joint = np.logical_and(cond1, cond2)
fig, axs = plt.subplots(2,1)
p = axs[0].plot(xc)
labels = ['unconditional', 'given x > 0 at T1', 'given x > 0.1 at T2', 'given x > 0 at T1 and x > 0.1 at T2']
plt.boxplot([xc[-1, :], xc[-1, cond1], xc[-1, cond2], xc[-1, cond_joint]], vert=False, labels=labels)
plt.show()
```
output looks like 
![image](https://user-images.githubusercontent.com/7674197/183820859-b8025e3a-d9f5-4e91-a1d1-10588ad5b29c.png)

so multiple conditioning isn't equivalent to single conditioning, at least in this really simple setup

Reply:
Ok, I get it now. Thanks everyone! I think my confusion stemmed from thinking about the allele frequency at the end of the rejection sampling process, rather than thinking about the *state of the simulation*.

So, this is a problem for all save/restore rejection sampling in SLiM then, right? If so, then this affects a lot of people and a lot of studies!

Reply:
It would be nice if we could convert this whole conversation to an issue, or a discussion. Anyone know if this is possible?

Reply:
@grahamgower It may be OK to checkpoint the state when a mutation is drawn, because in this case there's only one possible allele frequency (1/N) so nothing really would be influenced by the future conditioning. But, if multiple mutations are drawn on the same contig at different times, and are all checkpointed, then I think that would lead to the same issues as discussed above.

I'll go ahead and summarise the setup & explanations in this thread to post a new issue.

Reply:
I agree - to be extra clear, the problem occurs if there is a checkpointed state that is after the mutation has been introduced that might be returned to from a later point - in other words, if there's an event that has `save=True`, other than the initial mutation's introduction, followed by a rejection-sampling-based conditioning later in time.

And the reason that it *is* ok to do `save` at the mutation's introduction is that *that* is exactly what we want to condition on - the trajectory from that point forward. The way it works is that if we have `save` points at times `t0, t1, ..., tn`, then the trajectory is conditioned *separately* on each interval $[t_i, t_{i+1})$.

Reply:
Alrighty I opened an issue at https://github.com/popsim-consortium/stdpopsim/issues/1322. Maybe we should fix this in a separate PR, to keep the "single site string ID" as a separate addition ... I can see us going back-and-forth on some of the implementation details wrt to allowing checkpointing for `DrawMutation` events.